File: python/benchmark.py
"""Protocolo experimental do A* em Python.

Segue a metodologia da especificação do projeto:
- 3 cenários (melhor caso, caso médio, pior caso);
- tamanhos de entrada pequena/média/grande (+ intermediários para as curvas);
- 30 rodadas medidas por (cenário, tamanho), após 3 rodadas de aquecimento
  descartadas (estabiliza caches e, no caso do Node, o JIT — o mesmo
  protocolo é usado nas duas linguagens);
- medição com time.perf_counter_ns() (alta precisão), cronometrando APENAS a
  chamada do algoritmo (a geração da grade fica fora da medição);
- caso médio: cada rodada usa uma grade aleatória diferente (sementes 1..30,
  as MESMAS nas duas linguagens, graças ao PRNG compartilhado).

Saída: results/resultados_python.csv

Uso:
    python python/benchmark.py [--rapido]

    --rapido: 5 rodadas e só os tamanhos nomeados (para teste do script).
"""

from __future__ import annotations

import csv
import sys
import time
from pathlib import Path

from astar import astar
from scenarios import TAMANHOS, caso_medio, melhor_caso, pior_caso

LINGUAGEM = "python"
RODADAS = 30
AQUECIMENTO = 3
SEEDS_AQUECIMENTO = (9001, 9002, 9003)

RAIZ = Path(__file__).resolve().parent.parent
ARQUIVO_SAIDA = RAIZ / "results" / f"resultados_{LINGUAGEM}.csv"

CABECALHO = [
    "linguagem", "cenario", "rotulo", "lado", "n_celulas",
    "rodada", "seed", "tempo_ms", "nos_expandidos", "custo_caminho",
]


def cronometrar(grid, origem, destino):
    """Cronometra uma execução do A*. Retorna (tempo_ms, expandidos, custo)."""
    inicio = time.perf_counter_ns()
    _, custo, expandidos = astar(grid, origem, destino)
    fim = time.perf_counter_ns()
    return (fim - inicio) / 1e6, expandidos, custo


def executar(rodadas: int, tamanhos) -> list[list]:
    linhas_csv = []

    for cenario in ("melhor", "medio", "pior"):
        for lado, rotulo in tamanhos:
            n_celulas = lado * lado

            if cenario == "medio":
                # Aquecimento com grades que NÃO entram na medição.
                for seed in SEEDS_AQUECIMENTO[:AQUECIMENTO]:
                    grid, origem, destino, _ = caso_medio(lado, seed)
                    cronometrar(grid, origem, destino)
                for rodada in range(1, rodadas + 1):
                    grid, origem, destino, seed_usada = caso_medio(lado, rodada)
                    tempo_ms, expandidos, custo = cronometrar(grid, origem, destino)
                    linhas_csv.append([
                        LINGUAGEM, cenario, rotulo, lado, n_celulas,
                        rodada, seed_usada, f"{tempo_ms:.6f}", expandidos, custo,
                    ])
            else:
                gerador = melhor_caso if cenario == "melhor" else pior_caso
                grid, origem, destino = gerador(lado)
                for _ in range(AQUECIMENTO):
                    cronometrar(grid, origem, destino)
                for rodada in range(1, rodadas + 1):
                    tempo_ms, expandidos, custo = cronometrar(grid, origem, destino)
                    linhas_csv.append([
                        LINGUAGEM, cenario, rotulo, lado, n_celulas,
                        rodada, "", f"{tempo_ms:.6f}", expandidos, custo,
                    ])

            print(f"[{LINGUAGEM}] {cenario:<6} lado={lado:<4} ({rotulo}) ok",
                  file=sys.stderr)

    return linhas_csv


def main() -> None:
    rapido = "--rapido" in sys.argv
    rodadas = 5 if rapido else RODADAS
    tamanhos = [t for t in TAMANHOS if t[1] != "intermediaria"] if rapido else TAMANHOS

    linhas_csv = executar(rodadas, tamanhos)

    ARQUIVO_SAIDA.parent.mkdir(parents=True, exist_ok=True)
    with open(ARQUIVO_SAIDA, "w", newline="", encoding="utf-8") as f:
        escritor = csv.writer(f)
        escritor.writerow(CABECALHO)
        escritor.writerows(linhas_csv)

    print(f"{len(linhas_csv)} medições gravadas em {ARQUIVO_SAIDA}", file=sys.stderr)


if __name__ == "__main__":
    main()

File: javascript/benchmark.js
/**
 * Protocolo experimental do A* em JavaScript — espelho de python/benchmark.py.
 *
 * - 3 cenários (melhor, médio, pior) × tamanhos pequena/média/grande
 *   (+ intermediários), 30 rodadas medidas após 3 de aquecimento (importante
 *   no Node para estabilizar o JIT do V8);
 * - medição com process.hrtime.bigint() (nanossegundos), cronometrando
 *   APENAS a chamada do algoritmo;
 * - caso médio: sementes 1..30, as mesmas do Python (PRNG compartilhado),
 *   logo grades idênticas nas duas linguagens.
 *
 * Saída: results/resultados_javascript.csv
 *
 * Uso:
 *     node javascript/benchmark.js [--rapido]
 */

import { writeFileSync, mkdirSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import { astar } from "./astar.js";
import { TAMANHOS, casoMedio, melhorCaso, piorCaso } from "./scenarios.js";

const LINGUAGEM = "javascript";
const RODADAS = 30;
const AQUECIMENTO = 3;
const SEEDS_AQUECIMENTO = [9001, 9002, 9003];

const RAIZ = join(dirname(fileURLToPath(import.meta.url)), "..");
const ARQUIVO_SAIDA = join(RAIZ, "results", `resultados_${LINGUAGEM}.csv`);

const CABECALHO = [
  "linguagem", "cenario", "rotulo", "lado", "n_celulas",
  "rodada", "seed", "tempo_ms", "nos_expandidos", "custo_caminho",
];

/** Cronometra uma execução do A*. */
function cronometrar(grid, origem, destino) {
  const inicio = process.hrtime.bigint();
  const { custo, expandidos } = astar(grid, origem, destino);
  const fim = process.hrtime.bigint();
  return { tempoMs: Number(fim - inicio) / 1e6, expandidos, custo };
}

function executar(rodadas, tamanhos) {
  const linhasCsv = [];

  for (const cenario of ["melhor", "medio", "pior"]) {
    for (const [lado, rotulo] of tamanhos) {
      const nCelulas = lado * lado;

      if (cenario === "medio") {
        for (const seed of SEEDS_AQUECIMENTO.slice(0, AQUECIMENTO)) {
          const { grid, origem, destino } = casoMedio(lado, seed);
          cronometrar(grid, origem, destino);
        }
        for (let rodada = 1; rodada <= rodadas; rodada++) {
          const { grid, origem, destino, seedEfetiva } = casoMedio(lado, rodada);
          const { tempoMs, expandidos, custo } = cronometrar(grid, origem, destino);
          linhasCsv.push([
            LINGUAGEM, cenario, rotulo, lado, nCelulas,
            rodada, seedEfetiva, tempoMs.toFixed(6), expandidos, custo,
          ]);
        }
      } else {
        const gerador = cenario === "melhor" ? melhorCaso : piorCaso;
        const { grid, origem, destino } = gerador(lado);
        for (let i = 0; i < AQUECIMENTO; i++) cronometrar(grid, origem, destino);
        for (let rodada = 1; rodada <= rodadas; rodada++) {
          const { tempoMs, expandidos, custo } = cronometrar(grid, origem, destino);
          linhasCsv.push([
            LINGUAGEM, cenario, rotulo, lado, nCelulas,
            rodada, "", tempoMs.toFixed(6), expandidos, custo,
          ]);
        }
      }

      console.error(`[${LINGUAGEM}] ${cenario.padEnd(6)} lado=${String(lado).padEnd(4)} (${rotulo}) ok`);
    }
  }

  return linhasCsv;
}

function main() {
  const rapido = process.argv.includes("--rapido");
  const rodadas = rapido ? 5 : RODADAS;
  const tamanhos = rapido ? TAMANHOS.filter(([, r]) => r !== "intermediaria") : TAMANHOS;

  const linhasCsv = executar(rodadas, tamanhos);

  mkdirSync(join(RAIZ, "results"), { recursive: true });
  const conteudo = [CABECALHO.join(","), ...linhasCsv.map((l) => l.join(","))].join("\n") + "\n";
  writeFileSync(ARQUIVO_SAIDA, conteudo, "utf-8");

  console.error(`${linhasCsv.length} medições gravadas em ${ARQUIVO_SAIDA}`);
}

main();
